client/client.go

```diff
@@ -0,0 +1,479 @@
+	}
+}
+
+// refreshRESTMapper will call the discovery API and create a RESTMapper that is used to convert a GVK to a GVR. An
+// error is returned if the discovery API call failed.
+func (d *dynamicWatcher) refreshRESTMapper() error {
+	klog.V(2).Info("Refreshing the Kubernetes API REST Mapper")
+
+	discovery := d.client.Discovery()
+
+	apiGroups, err := restmapper.GetAPIGroupResources(discovery)
+	if err != nil {
+		klog.Error(err, "Could not get the API groups list from the Kubernetes API")
+
+		return err
+	}
+
+	d.restMapper = restmapper.NewDiscoveryRESTMapper(apiGroups)
+
+	return nil
+}
+
+// AddOrUpdateWatcher updates the watches for the watcher. When updating, any previously watched objects not specified
+// will stop being watched. If an error occurs, any created watches as part of this method execution will be removed.
+func (d *dynamicWatcher) AddOrUpdateWatcher(watcher ObjectIdentifier, watchedObjects ...ObjectIdentifier) error {
+	if !d.started {
+		return ErrNotStarted
+	}
+
+	if len(watchedObjects) == 0 {
+		return fmt.Errorf("%w: at least one watched object must be provided", ErrInvalidInput)
+	}
+
+	if err := watcher.Validate(); err != nil {
+		return err
+	}
+
+	for _, watchedObject := range watchedObjects {
+		if err := watchedObject.Validate(); err != nil {
+			return err
+		}
+	}
+
+	d.lock.Lock()
+	defer func() { d.lock.Unlock() }()
+
+	watchedObjectsSet := make(map[ObjectIdentifier]bool, len(watchedObjects))
+
+	var encounteredErr error
+
+	watchesAdded := []ObjectIdentifier{}
+
+	for _, watchedObject := range watchedObjects {
+		// If the object was previously watched, do nothing.
+		if d.watcherToWatches[watcher][watchedObject] {
+			watchedObjectsSet[watchedObject] = true
+
+			continue
+		}
+
+		if len(d.watchedToWatchers[watchedObject]) == 0 {
+			d.watchedToWatchers[watchedObject] = map[ObjectIdentifier]bool{}
+		}
+
+		// If the object is also watched another object, then do nothing.
+		if _, ok := d.watches[watchedObject]; ok {
+			watchedObjectsSet[watchedObject] = true
+			d.watchedToWatchers[watchedObject][watcher] = true
+
+			continue
+		}
+
+		gk := schema.GroupKind{Group: watchedObject.Group, Kind: watchedObject.Kind}
+
+		mapping, err := d.restMapper.RESTMapping(gk, watchedObject.Version)
+		if err != nil {
+			// Refresh the REST mapper and try again if the mapping wasn't found.
+			err = d.refreshRESTMapper()
+			if err != nil {
+				klog.Error(err, "failed to refresh the API discovery data")
+
+				encounteredErr = err
+
+				break
+			}
+
+			mapping, err = d.restMapper.RESTMapping(gk, watchedObject.Version)
+			if err != nil {
+				klog.Error(err, "Could not get resource mapping for %s", watchedObject)
+
+				encounteredErr = err
+
+				break
+			}
+		}
+
+		var resource dynamic.ResourceInterface = d.dynamicClient.Resource(mapping.Resource)
+		if watchedObject.Namespace != "" {
+			resource = resource.(dynamic.NamespaceableResourceInterface).Namespace(watchedObject.Namespace)
+		}
```

Comment:
Nice, one line smaller than what we've got in config-policy, I like it!

For reference: https://github.com/open-cluster-management-io/config-policy-controller/blob/08c00e374028867c071144585206dcea813a9097/controllers/configurationpolicy_controller.go#L557